packages/slider/src/HandleController.ts

```diff
@@ -0,0 +1,371 @@
+/*
+Copyright 2021 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+import { html, TemplateResult, ifDefined } from '@spectrum-web-components/base';
+import { streamingListener } from '@spectrum-web-components/base/src/streaming-listener.js';
+import { Slider } from './Slider';
+import { SliderHandle, HandleValues } from './SliderHandle';
+
+type HandleComponents = {
+    model: SliderHandle;
+    handle: HTMLElement;
+    input: HTMLInputElement;
+};
+
+export class HandleController {
+    private observer!: MutationObserver;
+    private host!: Slider;
+    private model: Map<string, SliderHandle> = new Map();
+    private handleOrder: string[] = [];
+    private boundingClientRect?: DOMRect;
+    private draggingHandle?: SliderHandle;
+
+    constructor(host: Slider) {
+        this.host = host;
+    }
+
+    public get values(): HandleValues {
+        const result: HandleValues = [];
+        for (const model of this.model.values()) {
+            result.push({
+                name: model.handleName,
+                value: model.value as number,
+            });
+        }
+        return result;
+    }
+
+    public set values(values: HandleValues) {
+        // TODO: Implement
+        console.log(values);
+    }
+
+    public get formattedValues(): Map<string, string> {
+        const result = new Map<string, string>();
+        for (const handle of this.model.values()) {
+            const formatOptions =
+                handle.formatOptions ?? this.host.formatOptions;
+            const formattedValue = formatOptions
+                ? new Intl.NumberFormat(
+                      navigator.language,
+                      formatOptions
+                  ).format(handle.value)
+                : handle.value.toString();
+            result.set(handle.handleName, formattedValue);
+        }
+        return result;
+    }
+
+    public get hasExplicitHandles(): boolean {
+        if (this.model.size === 0) return false;
+        if (this.model.size > 1) return true;
+        return this.model.values().next().value !== this.host;
```

Comment:
If the user has explicitly defined handles, then that changes how the `value` attribute on the `sp-slider` behaves.

This comes back to an earlier suggestion of yours that I had implemented. I have a `value` property on the `sp-slider` and `sp-slider-handle` that was always a `number`. I had then added a second property called `values` on the `sp-slider` that had the structured data for all of the sliders. You asked if I could just change the shape of the `value` property.

So, at the moment, there is only a `value` property and it returns a number when there are no explicit `sp-slider-handles`. When there are, it returns the structured `HandleValues` result. This is what is causing the break in the wild that you mentioned above...

I wonder if it would be better to go back to having two attributes? What do you think?

Reply:
It makes perfect sense later on in the review when I actually see it used, so really just a comment it fine for me.

Reply:
Comment added